### ISSUE TITLE
get http response when using stream API to upload file

### ISSUE THREAD

**Issue by leolumicrosoft** [sindresorhus/got#1514]
I can follow sample code to successfully upload file, like below
...
await pipeline(
		fs.createReadStream('index.html'),
		got.stream.post('https://sindresorhus.com')
	);
...
but I do not know how to get the response from http server, such as the status code and response body. Can you give me a sample code? Thanks

**Comment by szmarczak** [comment]
Duplicate of #771 and #774

**Comment by leolumicrosoft** [comment]
I can not figure out how 771 or 774 give me a solution. Can you directly give a brief sample code for my purpose? Note I am uploading file to a server and want to get the responds from server, NOT downloading.

What i tried,
await pipeline(
    fs.createReadStream('testindoormap.zip'),
    got.stream.post(uploadDrawPackageURL).on('response', response => {
      console.log("never reach this line of code")
      console.log(response.headers);
    }).resume()
  )

**Comment by szmarczak** [comment]
Hmmm. That sounds like a bug, the `response` event should be emitted. Can you create a RunKit example please? A repository works too.

**Comment by leolumicrosoft** [comment]
I have a small test repository for your try
https://github.com/leolumicrosoft/testGot.git

there are two JS, one is using GOT. The other is using request and this one gives server response. Please have a look. Thanks.

**Comment by Giotino** [comment]
We discusses this a lot in #1481 

TL; DR: If you care about the response from the server you need to read the response stream, here's a small example:

```js
const { pipeline, PassThrough } = require('stream');
const fs = require('fs');
const got = require('got');

const reader = new PassThrough();

const request = got.stream.post('https://postman-echo.com/post');
request.on('response', (response) => {
  console.log('now it reaches this line');
  console.log(response.headers);
});

request.resume();

pipeline(fs.createReadStream('testindoormap.zip'), request, reader, (err) => {
  console.log('HEY');
  if (err) {
    console.error(err);
  } else {
    console.log('done');
  }
});
```

**Comment by Giotino** [comment]
I updated the code in the previous comment. Previously I used a `stream.PassThrough` but it doesn't consume the stream, so the pipeline never end (it dies). With a `blackhole` (a Writable stream that only consumes the body) the pipeline terminate correcly.

If you want to read the body you can use a PassThrough but you've to read it, example:
```js
const { pipeline } = require('stream/promises');
const { PassThrough } = require('stream');
const fs = require('fs');
const got = require('got');
const getStream = require('get-stream');

const reader = new PassThrough();

const request = got.stream.post('http://127.0.0.1:8080/');
request.on('response', () => {
  console.log('response');
  (async () => {
    const body = await getStream(reader);
    console.log(body);
  })();
});

(async () => {
  await pipeline(
    fs.createReadStream('testindoormap.zip'),
    request,
    reader
  );

  console.log('Pipeline END');
})();
```

**Comment by Giotino** [comment]
I wrote the two example in order to give some code to explain how to handle streams, but what I think you (@leolumicrosoft) need is much simplier.
You can just:
```js
const fs = require('fs');
const got = require('got');

(async () => {
  const response = await got.post('http://127.0.0.1:8080/', {
    body: fs.createReadStream('testindoormap.zip')
  });

  console.log(response.headers);
  console.log(response.body);
})();
```

Because `options.body` can be a stream.

**Comment by leolumicrosoft** [comment]
Ok. I did not realize I do not need to use stream pipeline for uploading read stream, instead I can directly post with body as the read stream. It solves my problem, thanks.

**Comment by felixfbecker** [comment]
I still cannot figure out how to get the response body in stream mode. I cannot set `body` to a stream because I'm using [`new tar.Parse()`](https://www.npmjs.com/package/tar) which Got rejects as not a valid `Readable` stream (I'm guessing it does an `instanceof` check that fails?) but it can be piped fine, so I need to use `got.stream` to pipe the request body in. But I really just want a simple JSON body, and cannot figure out how to access the body at all in that mode. The example above with `new stream.PassThrough()` still didn't give me any body. `responseType` is also ignored per documentation.

**Comment by Giotino** [comment]
@felixfbecker I can't understand a single thing in your comment (you put there a lot of random things that make no sense), can you explain what are you trying to do?

If I really try to understand what's going on I think you tried to read the response body (a compressed JSON) by passing `tar.Parse()` as the request body.

I see a lot of wrong things here.
Why did you use the parser as the request body?
Why are you using tar for a compressed body? Tar is an archive of files, not a compression standard.
The example with `PassThrough` gave you no result because the entire point of that example was to discard the body (if you don't replace the `PassThrough` stream).

**Comment by szmarczak** [comment]
@felixfbecker See #1503 - your issue is almost the same.
